Improve build time

mundya opened the pull request:
Changes the way in which connections are stored so that the processing required when optimising out passthrough nodes is faster.

In addition this switches testing to test against the most recent release of Nengo (this may be changed back to master later but is up for discussion - @tcstewar).
## Details

Connections used to be stored using a list-derived structure, this is now changed to a dictionary. Since repeated hashing was found to be ludicrously slow a faster hashing package is used when installed. The changes in these commits simply change the data structure (1) and add support for this faster hashing (2). The change to the test requirements is (3).

mundya committed:
Use dictionary in connection map storage

This changes the format of the `ConnectionMap` data structure to use a
dictionary rather than a list of elements. As a result, fewer
comparisons are required when adding new connections to the map and this
leads to significant performance improvements when optimising out
passthrough nodes.

mundya committed:
Optionally use xxhash to accelerate network build

Reduces the time spent hashing Numpy arrays - consequently improves
network build time. If xxhash is not available md5 is used instead.

mundya committed:
Test alternative requirements

neworderofjamie commented on nengo_spinnaker/builder/connection.py:
```diff
@@ -43,12 +49,17 @@ def __eq__(self, other):
 
         return True
 
+    def _get_hashables(self):
```
@property ?

neworderofjamie commented on nengo_spinnaker/builder/model.py:
```diff
@@ -49,24 +51,11 @@ def add_connection(self, source_object, source_port, signal_parameters,
-            sinks = psp.sinks
-
-        # Add the parameters to the list of sinks
+        sinks = self._connections[source_object][source_port][pars]
```
Maybe add a note that the final `pars` is a hashable tuple of parameters (or something)

mundya replied:
See line 74 above, thanks

neworderofjamie commented on nengo_spinnaker/builder/model.py:
```diff
@@ -283,27 +281,26 @@ def remove_sinkless_signals(conn_map):
+
+        # Remove any parameter: sinks mappings where there are no sinks
+        for port, params_and_signals in iteritems(port_and_signals):
+            to_remove = [p for p, s in iteritems(params_and_signals) if not s]
```
`if s is None` ?

neworderofjamie commented on nengo_spinnaker/builder/model.py:
```diff
@@ -283,27 +281,26 @@ def remove_sinkless_signals(conn_map):
-                conns[source_object][source_port].append(parameters_and_sinks)
+            # If there is now nothing coming from this port then mark the port
+            # for removal.
+            if not params_and_signals:
```
`if len(params_and_signals) == 0` - not is obtuse to me

neworderofjamie commented on nengo_spinnaker/builder/model.py:
```diff
@@ -325,57 +322,46 @@ def remove_sinkless_objects(conn_map, cls):
+            transmitting_objects.add(source_object)
+
     # Identify all objects which receive but do not transmit
     remove_objects = receiving_objects - transmitting_objects
```
`-` is not set nand?

mundya committed:
Replace a _get_hashables(self) with a property

Somewhat neater, thanks @neworderofjamie

mundya committed:
Better document ConnectionMap

Improves the documentation and is a bit clearer about when ``not x``
really means ``len(x) == 0``.

Thanks @neworderofjamie

neworderofjamie reviewed (approved):
Looks good
